File: backend/app/schemas/soil.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class PhAssessment(BaseModel):
    ph: float
    ph_category: str
    is_valid: bool
    yield_multiplier: float


class SoilSuitabilityResponse(BaseModel):
    farm_id: int
    ph: PhAssessment
    nutrient_score: float
    fertility_category: str
    nitrogen: Optional[float] = None
    phosphorus: Optional[float] = None
    potassium: Optional[float] = None
    moisture: Optional[float] = None


class SoilReportCreate(BaseModel):
    farm_id: int
    ph: float = Field(ge=0, le=14)
    nitrogen: Optional[float] = Field(default=None, ge=0)
    phosphorus: Optional[float] = Field(default=None, ge=0)
    potassium: Optional[float] = Field(default=None, ge=0)
    moisture: Optional[float] = Field(default=None, ge=0)
    organic_carbon: Optional[float] = Field(default=None, ge=0)


class SoilReportUpdate(BaseModel):
    ph: Optional[float] = Field(default=None, ge=0, le=14)
    nitrogen: Optional[float] = Field(default=None, ge=0)
    phosphorus: Optional[float] = Field(default=None, ge=0)
    potassium: Optional[float] = Field(default=None, ge=0)
    moisture: Optional[float] = Field(default=None, ge=0)
    organic_carbon: Optional[float] = Field(default=None, ge=0)


class SoilReportResponse(BaseModel):
    """A single soil report plus every derived metric, computed fresh
    on read by SoilService — never persisted, so the formula can change
    without a backfill.
    """

    id: int
    farm_id: int
    farm_name: Optional[str] = None
    ph: float
    nitrogen: Optional[float] = None
    phosphorus: Optional[float] = None
    potassium: Optional[float] = None
    moisture: Optional[float] = None
    organic_carbon: Optional[float] = None
    ph_category: str
    nutrient_score: float
    fertility_category: str
    soil_health_index: float
    health_category: str
    created_at: datetime


class SoilReportListResponse(BaseModel):
    items: List[SoilReportResponse]
    total: int
    page: int
    page_size: int


class SoilDashboardStats(BaseModel):
    average_soil_health: Optional[float] = None
    healthy_farms: int
    poor_soil_farms: int
    total_farms_with_reports: int
    recent_reports: int
    total_reports: int

File: frontend/types/soil.ts
export interface SoilPh {
    ph: number;
    ph_category: string;
    is_valid: boolean;
    yield_multiplier: number;
}

export interface SoilSuitability {
    farm_id: number;
    ph: SoilPh;
    nutrient_score: number;
    fertility_category: string;
    nitrogen: number | null;
    phosphorus: number | null;
    potassium: number | null;
    moisture: number | null;
}

export interface SoilReport {
    id: number;
    farm_id: number;
    farm_name: string | null;
    ph: number;
    nitrogen: number | null;
    phosphorus: number | null;
    potassium: number | null;
    moisture: number | null;
    organic_carbon: number | null;
    ph_category: string;
    nutrient_score: number;
    fertility_category: string;
    soil_health_index: number;
    health_category: "healthy" | "moderate" | "poor";
    created_at: string;
}

export interface SoilReportListResponse {
    items: SoilReport[];
    total: number;
    page: number;
    page_size: number;
}

export interface SoilDashboardStats {
    average_soil_health: number | null;
    healthy_farms: number;
    poor_soil_farms: number;
    total_farms_with_reports: number;
    recent_reports: number;
    total_reports: number;
}

export interface SoilReportInput {
    farm_id: number;
    ph: number;
    nitrogen?: number;
    phosphorus?: number;
    potassium?: number;
    moisture?: number;
    organic_carbon?: number;
}
